### DISCUSSION TITLE
next js run script right after page load / nextjs 13 app / only works if there is an error / react

### DISCUSSION THREAD

**Discussion by romanusname** [vercel/next.js#44762] [Help]
I deliberately simplified the task to make it easier for you to understand.

_My goal_ - is when the page loads (useEffect) the JavaScript code changes the color of the text. I shouldn't be able to see the default color since useEffect runs instantly. **I have to instantly see blue**

```
'use client';
import React, { useEffect, useRef } from 'react';

  export default function index() {
  const ref = useRef();

  useEffect(() => {
    ref.current.style.color = 'blue'
  }, []);

  return <h1 ref={ref}>12345</h1>;
}
```
https://stackblitz.com/edit/nextjs-nzjwdb?file=app%2Fpage.js - example

_In practice_ - all JavaScript code in Next js runs with a delay. Because of this, I see the text in the default color for a second.
In an article with a similar problem, it was proposed to create a custom hook useIsomorphicLayoutEffect. My code should be like this:
```
'use client';
import React, { useRef,useEffect,useLayoutEffect } from 'react';

const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' ? useLayoutEffect : useEffect;

export default function index() {
  const ref = useRef();
  useIsomorphicLayoutEffect(() => {
    ref.current.style.color = 'blue'
  }, []);

  return <h1 ref={ref}>12345</h1>;
}
```
https://stackblitz.com/edit/nextjs-2dc5z4?file=app%2Fpage.js - example, custom hook, behavior unchanged
It didn't work. I still saw the default color at the beginning.

**Then I made a mistake!** If it's wrong to write "useEffect" in a custom hook, the script works as I want! When the page loads, the color is instantly blue!!!
```
const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' ? useLayoutEffect : useEffectBlaBlaBlaRandomText;
```
https://stackblitz.com/edit/nextjs-ofguvu?file=app%2Fpage.js - example, expected behavior with error in console
**I don't know why it works like this. The code works correctly only on error**
But I am getting an error in the console that "useEffectBlaBlaBlaRandomText is not defined" What can I do to make everything work and there are no errors in the console?

**Comment by GabenGar** [comment]
> I shouldn't be able to see the default color since useEffect runs instantly.

`useEffect()` doesn't run instantly, it only runs at best immediately after a first render.

> example, custom hook, behavior unchanged

In this small particular example it doesn't. But the [react docs](https://beta.reactjs.org/reference/react/useLayoutEffect) say this:

> The code inside useLayoutEffect and all state updates scheduled from it block the browser from repainting the screen. When used excessively, this can make your app very slow. When possible, prefer [useEffect.](https://beta.reactjs.org/reference/reac/useEffect)

So if you plan to use this custom hook as a drop-in replacement for `useEffect()`, your app will become progressively more janky, since every top level state changes will have to wait for all layout effects to run/rerun in all components in the subtree.

**Reply by romanusname** [reply]
Have an idea how to achieve the desired result? I really need help
